File: tools/art-pipeline/build-runtime-artwork-atlases.swift
import CoreGraphics
import Foundation
import ImageIO
import UniformTypeIdentifiers

struct AtlasLevel: Codable {
    let x: Int
    let y: Int
    let w: Int
    let h: Int
}

struct AtlasEntry: Codable {
    let assetId: String
    let url: String
    let width: Int
    let height: Int
    let levels: [String: AtlasLevel]
}

struct AtlasManifest: Codable {
    let schemaVersion: String
    let generatedBy: String
    let sourceRoot: String
    let outputRoot: String
    let tileSize: Int
    let columns: Int
    let rows: Int
    let fileCount: Int
    let entries: [AtlasEntry]
}

let arguments = CommandLine.arguments
func argument(_ name: String, fallback: String) -> String {
    guard let index = arguments.firstIndex(of: name), index + 1 < arguments.count else { return fallback }
    return arguments[index + 1]
}

let sourceRoot = URL(fileURLWithPath: argument("--source", fallback: "public/assets/buildings-runtime-384"), isDirectory: true)
let outputRoot = URL(fileURLWithPath: argument("--out", fallback: "public/assets/buildings-runtime-atlas"), isDirectory: true)
let tileSize = Int(argument("--size", fallback: "384")) ?? 384
let columns = 3
let rows = 3
let fileManager = FileManager.default
try fileManager.createDirectory(at: outputRoot, withIntermediateDirectories: true)

let directories = try fileManager.contentsOfDirectory(at: sourceRoot, includingPropertiesForKeys: [.isDirectoryKey], options: [.skipsHiddenFiles])
    .filter { (try? $0.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true }
    .sorted { $0.lastPathComponent < $1.lastPathComponent }

let colorSpace = CGColorSpaceCreateDeviceRGB()
var entries: [AtlasEntry] = []

for directory in directories {
    var images: [Int: CGImage] = [:]
    for level in 0..<9 {
        let url = directory.appendingPathComponent("level-\(level).png")
        guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
              let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
            throw NSError(domain: "LittleEarAtlas", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing or unreadable \(url.path)"])
        }
        images[level] = image
    }

    guard let context = CGContext(
        data: nil,
        width: tileSize * columns,
        height: tileSize * rows,
        bitsPerComponent: 8,
        bytesPerRow: tileSize * columns * 4,
        space: colorSpace,
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue,
    ) else { throw NSError(domain: "LittleEarAtlas", code: 2) }
    context.clear(CGRect(x: 0, y: 0, width: tileSize * columns, height: tileSize * rows))

    var levels: [String: AtlasLevel] = [:]
    for level in 0..<9 {
        let column = level % columns
        let row = level / columns
        // CGContext's origin is bottom-left; manifest coordinates are top-left.
        let drawRect = CGRect(x: column * tileSize, y: (rows - 1 - row) * tileSize, width: tileSize, height: tileSize)
        if let image = images[level] { context.draw(image, in: drawRect) }
        levels[String(level)] = AtlasLevel(x: column * tileSize, y: row * tileSize, w: tileSize, h: tileSize)
    }

    let atlasURL = outputRoot.appendingPathComponent("\(directory.lastPathComponent).png")
    guard let destination = CGImageDestinationCreateWithURL(atlasURL as CFURL, UTType.png.identifier as CFString, 1, nil),
          let atlasImage = context.makeImage() else { throw NSError(domain: "LittleEarAtlas", code: 3) }
    CGImageDestinationAddImage(destination, atlasImage, nil)
    guard CGImageDestinationFinalize(destination) else { throw NSError(domain: "LittleEarAtlas", code: 4) }
    entries.append(AtlasEntry(
        assetId: directory.lastPathComponent,
        url: "/assets/buildings-runtime-atlas/\(directory.lastPathComponent).png",
        width: tileSize * columns,
        height: tileSize * rows,
        levels: levels,
    ))
}

let manifest = AtlasManifest(
    schemaVersion: "runtime-artwork-atlas.v1",
    generatedBy: "tools/art-pipeline/build-runtime-artwork-atlases.swift",
    sourceRoot: sourceRoot.path,
    outputRoot: outputRoot.path,
    tileSize: tileSize,
    columns: columns,
    rows: rows,
    fileCount: entries.count,
    entries: entries,
)
let encoder = JSONEncoder()
encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
try encoder.encode(manifest).write(to: outputRoot.appendingPathComponent("runtime-artwork-atlas-manifest.json"))
print("Runtime artwork atlases: \(entries.count) files, 252 frames -> \(outputRoot.path)")

File: tools/art-pipeline/build-runtime-artwork-atlases.js
#!/usr/bin/env node

import { mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from 'node:fs'
import { join, resolve } from 'node:path'
import { chromium } from 'playwright'

const sourceRoot = resolve(process.env.ATLAS_SOURCE ?? 'public/assets/buildings-runtime-384')
const outputRoot = resolve(process.env.ATLAS_OUT ?? 'public/assets/buildings-runtime-atlas-webp')
const tileSize = Number.parseInt(process.env.ATLAS_TILE_SIZE ?? '384', 10)
const format = process.env.ATLAS_FORMAT === 'png' ? 'png' : 'webp'
const quality = Math.min(1, Math.max(0.5, Number.parseFloat(process.env.ATLAS_QUALITY ?? '0.9')))
const columns = 3
const rows = 3
if (!Number.isInteger(tileSize) || tileSize < 64) throw new Error(`Invalid ATLAS_TILE_SIZE: ${tileSize}`)
mkdirSync(outputRoot, { recursive: true })

const assetIds = readdirSync(sourceRoot)
  .filter((entry) => statSync(join(sourceRoot, entry)).isDirectory())
  .sort()
if (assetIds.length === 0) throw new Error(`No runtime artwork directories under ${sourceRoot}`)

const browser = await chromium.launch({ headless: true })
const page = await browser.newPage()
const entries = []
try {
  for (const assetId of assetIds) {
    const images = Array.from({ length: 9 }, (_, level) => {
      const file = join(sourceRoot, assetId, `level-${level}.png`)
      return `data:image/png;base64,${readFileSync(file).toString('base64')}`
    })
    const dataUrl = await page.evaluate(async ({ images, tileSize, columns, rows, format, quality }) => {
      const canvas = document.createElement('canvas')
      canvas.width = tileSize * columns
      canvas.height = tileSize * rows
      const context = canvas.getContext('2d')
      if (!context) throw new Error('Canvas 2D context unavailable')
      context.clearRect(0, 0, canvas.width, canvas.height)
      await Promise.all(images.map((src, level) => new Promise((resolve, reject) => {
        const image = new Image()
        image.onload = () => {
          context.drawImage(image, (level % columns) * tileSize, Math.floor(level / columns) * tileSize, tileSize, tileSize)
          resolve()
        }
        image.onerror = reject
        image.src = src
      })))
      return canvas.toDataURL(`image/${format}`, quality)
    }, { images, tileSize, columns, rows, format, quality })
    const extension = format === 'webp' ? 'webp' : 'png'
    const output = join(outputRoot, `${assetId}.${extension}`)
    const prefix = `data:image/${format};base64,`
    if (!dataUrl.startsWith(prefix)) throw new Error(`Browser cannot encode ${format} atlas.`)
    writeFileSync(output, Buffer.from(dataUrl.slice(prefix.length), 'base64'))
    const levels = Object.fromEntries(Array.from({ length: 9 }, (_, level) => [String(level), {
      x: (level % columns) * tileSize,
      y: Math.floor(level / columns) * tileSize,
      w: tileSize,
      h: tileSize,
    }]))
    entries.push({
      assetId,
      url: `/assets/${outputRoot.split('/assets/').at(-1)}/${assetId}.${extension}`,
      width: tileSize * columns,
      height: tileSize * rows,
      levels,
    })
  }
} finally {
  await browser.close()
}

writeFileSync(join(outputRoot, 'runtime-artwork-atlas-manifest.json'), `${JSON.stringify({
  schemaVersion: 'runtime-artwork-atlas.v1',
  generatedBy: 'tools/art-pipeline/build-runtime-artwork-atlases.js',
  sourceRoot: 'public/assets/buildings-runtime-384',
  outputRoot: 'public/assets/buildings-runtime-atlas-webp',
  tileSize,
  columns,
  rows,
  fileCount: entries.length,
  frameCount: entries.length * 9,
  format,
  quality,
  entries,
}, null, 2)}\n`)
console.log(`Runtime artwork atlases: ${entries.length} files, ${entries.length * 9} frames -> ${outputRoot}`)
